### PR TITLE
bundle nltk tiktoken v2

### PR DESCRIPTION
**Pull request by logan-markewich**
# Description

The second approach at bundling nltk and tiktoken with the package.

Instead of using LFS (which has actual $$ costs associated with it), we can use a similar approach that will download and cache the tiktoken and nltk packages during publishing

Locally, I tested this with `poetry publish` -- you can try `pip install 0.9.16.dev2` and inspect the installed files -- they will include the cached files in `site_packages/llama_index/llama_index/_static` 

I'm not 100% sure how to test the github workflow without making an actual release, but I think it should work? 🙏🏻 

Fixes https://github.com/run-llama/llama_index/issues/9610

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -10,24 +10,30 @@ repos:
...
         args: [--fix, --exit-non-zero-on-fix]
+        exclude: llama_index/_static
   - repo: https://github.com/psf/black-pre-commit-mirror
     rev: 23.10.1
     hooks:
       - id: black-jupyter
         name: black-src
         alias: black
-        exclude: docs/
+        exclude: ^(docs/|llama_index/_static)
```
Comment (hatianzhang) [inline .pre-commit-config.yaml:36]: is this only excluding `docs/|llama_index/_static)`

Comment (logan-markewich) [reply .pre-commit-config.yaml:36]: it's a regex, exclude `docs/` or `llama_index/_static` from being run with `black`

File: llama_index/indices/keyword_table/utils.py [hunk 2]
```diff
@@ -29,16 +29,17 @@ def rake_extract_keywords(
     """Extract keywords with RAKE."""
     try:
         import nltk
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/indices/keyword_table/utils.py:31]: yea true

[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/indices/keyword_table/utils.py:31]: Actually we need the import, it's used later on.

**Review by hatianzhang — APPROVED** [review]
LGTM!

**Comment by cbornet** [comment]
IIUC, tiktoken and NLTK caches are downloaded at publish time.
I think the NLTK packages are downloaded from GitHub's main. So this means the build is not reproducible as the content of the NLTK package can evolve.

**Comment by logan-markewich** [comment]
@cbornet these files are not updated often (if at all), so I'm not really that co concerned.

If you have a better approach, I'm open to a PR. But as it stands this will greatly improve the YX for anyone running in an offline environment

**Comment by cbornet** [comment]
A better approach would be to download a fix version of the packages. So not using the NLTK downloader.
Eg: https://github.com/nltk/nltk_data/blob/1d3c34b4cfd6059986bf4bc604e5929335ab92ff/packages/tokenizers/punkt.zip

**Pull request merged**